feat: changes

luzzif opened the pull request.

Commit:
feat: enable cors on specification validation api

Commit:
chore(docker-compose): create defillama-answerer db by default in postgres container

Commit:
chore: update default config example

Commit:
feat: partially rename snapshots table to checkpoints

Commit:
fix: fix to/from sql implementation for address and specification

Commit:
fix: fix ipfs specification fetching by using post instead of get in cat call

Commit:
fix: do multiple calls instead of only one to get oracle data while parsing kpi token creation logs

Commit:
fix: correctly print anyhow errors with context instead of truncating that

Commit:
feat: add first integration tests

Commit:
fix: use specific chain id when creating new signer middleware

Commit:
feat: changes

- turn defillama module in a defillama client struct in order to make it easier to mock defillama http calls
- use arc to send defillama client around the app
- first proper implementation of answering logic

Commit:
fix: fix test context in integration tests to have 2 connections (one to the maintenance db and one to the test db)

Commit:
chore: start postgres service before executing tests in gh actions

guerrap commented on .github/workflows/ci.yml:
```diff
@@ -9,6 +9,14 @@ on:
+        image: postgres:latest
+        ports:
+          - 5432:5432
+        env:
```
I think we should add the `POSTGRES_DB` here, no? So we can define the default database name

luzzif replied:
This is purposefully left blank as this service is used while executing integration tests with the database. Each test accesses the db by instantiating the `TestContext` struct, that automatically connects to the default (always available) `postgres` database and creates a test-specific database for each run. Then, thanks to a custom `drop` implementation on the strcut, the database is deleted after the context goes out of scope (i.e. when the test is done). That's why we don't need a specific db name. It would just create a database that we wouldn't use ion this setting.

guerrap commented on Cargo.toml:
```diff
@@ -42,4 +43,6 @@ anyhow = "1.0.71"
 [dev-dependencies]
-mockito = "1.1.0"
+diesel = { version = "2.1.0", features = ["postgres", "r2d2", "serde_json"] }
+diesel_migrations = { version = "2.1.0", features = ["postgres"] }
```
Is it desired to have `diesel` and `diesel_migrations` both in `[dependencies]` and `[dev-dependencies]`?

luzzif replied:
Yes as we use it in both the app and in the integration tests that interact with the database.

guerrap commented on src/defillama.rs:
```diff
@@ -1,24 +1,42 @@
+            .send()
+            .await
+            .context(format!(
+                "could not get currentl tvl for protocol {}",
```
Small typo in `currentl`.

luzzif replied:
Addressed in cf5c3bacff9bf0051b81a0250298f01f81c049b0

guerrap commented on src/ipfs.rs:
```diff
@@ -17,14 +17,14 @@ pub async fn fetch_specification_with_retry(
     let fetch = || async {
         let response = match ipfs_http_client
-            .request(Method::GET, format!("/api/v0/get?arg={cid}"))?
+            .request(Method::POST, format!("/api/v0/cat?arg={cid}"))?
```
Why the API change here?

luzzif replied:
Well the method change is because we're actually interacting with a JSON RPC API, so POST must always be used. As for the actual command change, `get` apparently returns more stuff than simply the content of the file, so the JSON deserialization wasn't working good

guerrap commented on src/scanner/present.rs:
```diff
@@ -171,44 +175,134 @@ async fn handle_active_oracles_answering(
+
+    // wait forever unless some task stops with an error
+    while let Some(res) = join_set.join_next().await {
+        let _ = res.context("task unexpectedly stopped")?;
```
Is the `let _ =` here needed?

luzzif replied:
Yessir, otherwise we have a warning regarding the `Result` not being used

Commit:
chore: fix typo

Commit:
feat: improvements anbd fixes

- add logs and improve logging in general
- switch order of spawning of present and past indexers to avoid errors
  in communication if the past indexer has nothing to do but the present
  indexer hasn't started yet
- in the checkpoint model, make get_for_chain_id return a result of an
  option. not found errors are caught and none is returned, while if a
  generic error happens, it is broadcast to the caller. if a checkpoint
  is found for the chain id it's just returned
- implement basic integration tests for the checkpoint model

guerrap commented on src/scanner/past.rs:
```diff
@@ -26,29 +26,32 @@ pub async fn scan<'a>(
+                // realistically, the following should never happen
+                u64::try_from(checkpoint.block_number).expect(
+                    format!(
+                        "could not convert checkpoint block number {} to usigned integer",
```
Small type here `usigned`.

luzzif replied:
Addressed in 5e71d3f3f6bc58bc623f5004acd7070f77955ac1

Commit:
chore: fix typo
